### ISSUE TITLE
iOS and Android app freeze after authenticate with Facebook and API call

### ISSUE THREAD

**Issue by Mukarillo** [firebase/quickstart-unity#648]
Unity editor version: **2019.3.9f1**
Firebase Unity SDK version: **6.13.0**
Source you installed the SDK (.unitypackage or Unity Package Manager): **.unitypackage**
Firebase plugins in use (Auth, Database, etc.): **Auth, Crashlitycs, Messaging, Storage, Remote Config**
Additional SDKs you are using (Facebook, AdMob, etc.): **Facebook, Google Sign in**
Platform you are using the Unity editor on (Mac, Windows, or Linux): **Windows and Mac**
Platform you are targeting (iOS, Android, and/or desktop): **Android and iOS**
Scripting Runtime (Mono, and/or IL2CPP): **IL2CPP**
API Compatibility: **.NET 4.x**

### Please describe the issue here:
Last week I updated Firebase sdk from 6.8 to 6.13 and facebook sdk from 7.18.1 to 7.19.2. The game runs as expected in Editor, I can log in, interact with the games API that stores and collect data from Firebase database, authenticate with facebook and email/password and everything else.
Yesterday I made the first build after updating the sdks and the game is freezing after making the first api call when I am authenticated with facebook. This does not happens when I authenticate with Google or email/password. Also, if you authenticate with one of those and then while playing the game, log in with facebook, it works properly.

To Sum up:
1. You open the game, authenticate with facebook, make an api call and game freezes.
2. You open the game, authenticate with email/password or Google, make an api call, game works perfectly.

The problem is that there is no error, warning or anything else. The app simple stop spitting logs in logcat, so it's really hard to debug.

I am attaching the logcat file here.
[com.vrtron.projectword-logcat.txt](https://github.com/firebase/quickstart-unity/files/4486940/com.vrtron.projectword-logcat.txt)

Also, after a while freezed, both android and iOS kills the application.

### Please answer the following, if applicable:
Have you been able to reproduce this issue with just the Firebase Unity quickstarts (this GitHub project)?
I will try late today and update here.

What's the issue repro rate? 
100% in iOS and Android
0% in editor

**Comment by patm1987** [comment]
Hi @Mukarillo , thanks for your report!

There are a lot of SDKs setup in your project. If at all possible, it would help if you could reproduce this issue in the Firebase quickstart projects (probably the [auth one](https://github.com/firebase/quickstart-unity/tree/master/auth/testapp))?

Without that, you'll have to do a lot of the legwork to do a lot of testing and reporting.

Since you're talking about freezing, I would suggest taking a look at a couple of possibilities in your logic. Namely, I usually ask people to replace `ContinueWith` with [ContinueWithOnMainThread](https://firebase.google.com/docs/reference/unity/class/firebase/extensions/task-extension) whenever possible. When issues appear related to these, they're typically poorly reported.

Next, I see that you are using Coroutines at least a little bit from your logging. If you can, I'd recommend that you integrate Crashlytics (Crashlytics is free, but I understand if it wouldn't be compatible with your privacy policy). I cover debugging Coroutines in [this article](https://medium.com/firebase-developers/crashlytics-and-unity-part-1-ca0598862918).

My thought process behind this recommendation is that a Coroutine will often silently fail if an Exception is raised in it. Crashlytics on Unity hooks into C#'s exception reporting system, so even non-crash exceptions will light up in the dashboard. Reporting is also fast, you can often cause an exception and see it in the Crashlytics console within a few minutes.

So to recap:
Please repro in the quickstart, it helps out the team a ton.
Check for `ContinueWith`, replace with `ContinueWithOnMainThread` for testing.
Crashlytics might shed light on any exceptions that are causing `Coroutines` to quit

Let me know if any of that helps and what you find!
--Patrick

**Comment by Mukarillo** [comment]
Hello @patm1987 

Thank you very much for your reply, it helped me to narrow the search for the problem.
All my calls were already using `ContinueWithOnMainThread` and the game was working fine on Editor, so I knew that was not the problem, but that was a good advice.
I had Crashlytics in my project and after reading your comment, I checked the panel just to find some weird stack in iOS(maybe this makes sense to you) and nothing on Android:

<details>
<summary>main-thread stack</summary>
<code style="white-space:nowrap;">
Crashed: com.apple.main-thread
0  UnityFramework                 0x106c72474 UnityIsOrientationEnabled + 81740
1  UnityFramework                 0x106043308 EnabledAutorotationInterfaceOrientations() + 178 (UnityViewControllerBase+iOS.mm:178)
2  UnityFramework                 0x106037fc8 -[UnityAppController application:supportedInterfaceOrientationsForWindow:] + 201 (UnityAppController.mm:201)
3  UIKitCore                      0x1b5948140 -[UIApplication _supportedInterfaceOrientationsForWindow:] + 108
4  UIKitCore                      0x1b5343a64 -[UIViewController __supportedInterfaceOrientations] + 256
5  UIKitCore                      0x1b5335028 -[UIViewController _preferredInterfaceOrientationGivenStatusBarAndDeviceAndOrientation:] + 104
6  UIKitCore                      0x1b5343680 -[UIViewController _preferredInterfaceOrientationForPresentationInWindow:fromInterfaceOrientation:] + 544
7  UIKitCore                      0x1b52541c8 -[_UIFullscreenPresentationController _adjustOrientationIfNecessaryInWindow:forViewController:preservingViewController:] + 576
8  UIKitCore                      0x1b5249eb8 -[UIPresentationController _dismissWithAnimationController:interactionController:target:didEndSelector:] + 684
9  UIKitCore                      0x1b533e01c __99-[UIViewController _dismissViewControllerWithAnimationController:interactionController:completion:]_block_invoke_2 + 72
10 UIKitCore                      0x1b5dccc14 +[UIView(Animation) performWithoutAnimation:] + 96
11 UIKitCore                      0x1b533de98 -[UIViewController _dismissViewControllerWithAnimationController:interactionController:completion:] + 580
12 UIKitCore                      0x1b5dccc14 +[UIView(Animation) performWithoutAnimation:] + 96
13 UIKitCore                      0x1b533db4c -[UIViewController _dismissViewControllerWithTransition:from:completion:] + 816
14 UIKitCore                      0x1b533d228 -[UIViewController dismissViewControllerWithTransition:completion:] + 1200
15 UIKitCore                      0x1b533cc90 -[UIViewController _performCoordinatedPresentOrDismiss:animated:] + 508
16 UIKitCore                      0x1b533f3a0 -[UIViewController dismissViewControllerAnimated:completion:] + 132
17 UIKitCore                      0x1b597c28c __68-[UIWindow _removeAllViewControllersFromWindowHierarchyFromDealloc:]_block_invoke + 60
18 UIKitCore                      0x1b5337154 +[UIViewController _performWithoutDeferringTransitionsAllowingAnimation:actions:] + 168
19 UIKitCore                      0x1b597bf1c -[UIWindow _removeAllViewControllersFromWindowHierarchyFromDealloc:] + 288
20 UIKitCore                      0x1b597c438 -[UIWindow setRootViewController:] + 284
21 UnityFramework                 0x106037eac UnityCleanupTrampoline + 168 (UnityAppController.mm:168)
22 UnityFramework                 0x106038c68 -[UnityAppController applicationWillTerminate:] + 468 (UnityAppController.mm:468)
23 UIKitCore                      0x1b5949ae0 -[UIApplication _terminateWithStatus:] + 244
24 UIKitCore                      0x1b510f8b8 -[_UISceneLifecycleMultiplexer _evalTransitionToSettings:fromSettings:forceExit:withTransitionStore:] + 124
25 UIKitCore                      0x1b510f538 -[_UISceneLifecycleMultiplexer forceExitWithTransitionContext:scene:] + 216
26 UIKitCore                      0x1b593fe80 -[UIApplication workspaceShouldExit:withTransitionContext:] + 212
27 FrontBoardServices             0x1b69ea90c -[FBSUIApplicationWorkspaceShim workspaceShouldExit:withTransitionContext:] + 84
28 FrontBoardServices             0x1b6a1605c __83-[FBSWorkspaceScenesClient willTerminateWithTransitionContext:withAcknowledgement:]_block_invoke_2 + 76
29 FrontBoardServices             0x1b69fbf04 -[FBSWorkspace _calloutQueue_executeCalloutFromSource:withBlock:] + 232
30 FrontBoardServices             0x1b6a15fec __83-[FBSWorkspaceScenesClient willTerminateWithTransitionContext:withAcknowledgement:]_block_invoke + 124
31 libdispatch.dylib              0x1b158b184 _dispatch_client_callout + 16
32 libdispatch.dylib              0x1b1533fd8 _dispatch_block_invoke_direct$VARIANT$mp + 224
33 FrontBoardServices             0x1b6a3b418 __FBSSERIALQUEUE_IS_CALLING_OUT_TO_A_BLOCK__ + 40
34 FrontBoardServices             0x1b6a3b0e4 -[FBSSerialQueue _queue_performNextIfPossible] + 404
35 FrontBoardServices             0x1b6a3b60c -[FBSSerialQueue _performNextFromRunLoopSource] + 28
36 CoreFoundation                 0x1b183ca00 __CFRUNLOOP_IS_CALLING_OUT_TO_A_SOURCE0_PERFORM_FUNCTION__ + 24
37 CoreFoundation                 0x1b183c958 __CFRunLoopDoSource0 + 80
38 CoreFoundation                 0x1b183c0f0 __CFRunLoopDoSources0 + 180
39 CoreFoundation                 0x1b183723c __CFRunLoopRun + 1080
40 CoreFoundation                 0x1b1836adc CFRunLoopRunSpecific + 464
41 GraphicsServices               0x1bb7d7328 GSEventRunModal + 104
42 UIKitCore                      0x1b594463c UIApplicationMain + 1936
43 UnityFramework                 0x10603d1ec -[UnityFramework runUIApplicationMainWithArgc:argv:] + 96 (main.mm:96)
44 projectword                    0x104b37e1c main + 28 (main.mm:28)
45 libdyld.dylib                  0x1b16c0360 start + 4
</code>
</details>

I decided to reproduce the problem in Firebase Unity quickstarts and after a while setting the project up, the problem was not there. So it was something from my implementation.

I took the hardway and placed a bunch of comments on my code and managed to narrow down to a method requesting for facebook friends information in Firebase that I was calling as soon as I got logged in in facebook.
In the end, I am not 100% sure what was happening, I moved this call to when the game is fully loaded and now its working again. Maybe I haven't received all the friends ids from facebook, or I wasn't authenticated with firebase at that time. The weird part is that it works in Editor.

Thank you very much for the fast response and the detailed explanation. That was really nice!

**Comment by Mukarillo** [comment]
Hello @DellaBitta 

So, as soon as I got the callback from facebook that the user was logged in and finished getting the users friends ids, I was calling a firebase function to get the score from these players. At this moment the app simply hang and do not log anything.

I think it was some kind of race condition, or something.

**Comment by DellaBitta** [comment]
Ok, great, and thanks for following up with those details.
